File: output_repair/qwen3/sample_3/repair_3/c/c/Struct/deposit.c
#include "stdlib.h"
#include "assert.h"

struct account
{
    int balance;
};

/*@ predicate account(struct account *a; int b) = a->balance |-> b; @*/

//@ requires true
//@ ensures account(result, 0)
struct account *create_account()
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    myAccount->balance = 0;
    return myAccount;
}

//@ requires account(myAccount, ?old_balance)
//@ ensures account(myAccount, old_balance + amount)
void account_deposit(struct account *myAccount, int amount)
{
    myAccount->balance += amount;
}

//@ requires account(myAccount, ?b)
//@ ensures account(myAccount, b) &*& result == b
int account_get_balance(struct account *myAccount)
{
    int result = myAccount->balance;
    return result;
}

//@ requires account(myAccount, ?old_balance)
//@ ensures account(myAccount, newBalance)
void account_set_balance(struct account *myAccount, int newBalance)
{
    myAccount->balance = newBalance;
}

//@ requires account(myAccount, ?b)
//@ ensures true
void account_dispose(struct account *myAccount)
{
    free(myAccount);
}

//@ requires true
//@ ensures true
int main()
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_deposit(myAccount, 10);
    int b = account_get_balance(myAccount);
    assert(b == 15);
    account_dispose(myAccount);
    return 0;
}

File: output/deepseek/rust/purely_unsafe/solutions/deposit.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

struct Account {
    balance: i32,
}

predicate account(my_account: *mut Account, balance: i32) {
    my_account != 0 && struct_Account_balance(my_account, balance)
}

unsafe impl Account {
    #[requires(layout.size() == std::mem::size_of::<Account>())]
    #[requires(layout.align() == std::mem::align_of::<Account>())]
    #[ensures(result != 0)]
    #[ensures(account(result, 0))]
    unsafe fn create() -> *mut Account {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        (*my_account).balance = 0;
        return my_account;
    }

    #[requires(account(my_account, balance))]
    #[ensures(account(my_account, balance))]
    #[ensures(result == balance)]
    unsafe fn get_balance(my_account: *mut Account) -> i32 {
        return (*my_account).balance;
    }

    #[requires(account(my_account, _old_balance))]
    #[ensures(account(my_account, new_balance))]
    unsafe fn set_balance(my_account: *mut Account, new_balance: i32) {
        (*my_account).balance = new_balance;
    }

    #[requires(account(my_account, old_balance))]
    #[ensures(account(my_account, old_balance + amount))]
    unsafe fn deposit(my_account: *mut Account, amount: i32) {
        (*my_account).balance += amount;
    }

    #[requires(account(my_account, _balance))]
    #[ensures(true)]
    unsafe fn dispose(my_account: *mut Account) {
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }
}

fn main() {
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::deposit(my_account, 10);
        let b = Account::get_balance(my_account);
        std::hint::assert_unchecked(b == 15);
        Account::dispose(my_account);
    }
}